File: frontend/src/types/blockWorld.ts
export interface BlockGameState {
  stacks: string[][];
  holding: string | null;
  initial_stacks: string[][];
  goal_stacks: string[][];
  moves_count: number;
  is_solved: boolean;
}

export interface BlockSolveResponse {
  moves: string[];
  solution_states: Array<{
    stacks: string[][];
    holding: string | null;
  }>;
  explanation: string[];
}

File: backend/app/schemas/block_world.py
from typing import Optional
from pydantic import BaseModel


class BlockMoveRequest(BaseModel):
    action: str  # "pickup", "putdown", "stack", "unstack"
    block: str
    target: Optional[str] = None  # target block or table stack index


class BlockGameState(BaseModel):
    stacks: list[list[str]]
    holding: Optional[str] = None
    goal_stacks: list[list[str]]
    moves_count: int = 0
    is_solved: bool = False


class BlockSolveResponse(BaseModel):
    moves: list[str]
    solution_states: list[dict]
    explanation: list[str]
